MIT-licensed SparseMatrixCSC permute[!] and refactored [c]transpose[!]

Sacha0 opened the pull request:
Followup to #13001, #14631, #14702, #14798, and #15242. This pull request refactors `[c]transpsose[!]` methods operating on `SparseMatrixCSC`s and introduces accompanying MIT-licensed `permute[!]` methods as a replacement for `csc_permute`. Details:

The first commit refactors the `[c]transpose[!]` methods: (1) such that the later-introduced `permute[!]` methods share the `[c]tranpose[!]` methods' code insofar as possible; and (2) taking advantage of #16371 for better structuring. Part of (1) is replacing `qftranspose!` with a more general `halfperm!`; in that replacement, the permutation vector convention changes to match that seemingly used elsewhere in Julia (such that `halfperm!(X, A, q)` now produces `X = transpose(A[:,q])` and consequently `permute(A, p, q)` produces `A[p,q]` --- desired behavior?). This first commit also adds several tests for those `[c]tranpsose[!]` methods.

Would one of the powers that be ask NanoSoldier to run the sparse performance tests to check for regressions in `[c]transpose[!]`?

The second commit introduces `permute[!]` methods operating on `SparseMatrixCSC`s (intended to replace and extend the functionality of `csc_permute`), documents those methods, and adds a set of associated tests. In anticipation of heap-referencing immutables one day being stack-allocatable, the present interface operates solely on `SparseMatrixCSC`s rather than also providing methods working on sets of vectors constituting `SparseMatrixCSC`s. The second commit exports `permute` from `SparseArrays` but does not place an entry in `exports.jl` (should it?).

 Concerning `permute[!]`'s performance, these benchmarks

``` julia
using Benchmarks: @benchmark, SummaryStatistics
import Base.SparseArrays.permute
import Base.SparseArrays.csc_permute

function prettytimes(bench)
    stats = SummaryStatistics(bench)
    timecenter = stats.elapsed_time_center
    timelower = get(stats.elapsed_time_lower)
    timeupper = get(stats.elapsed_time_upper)
    # based on Benchmarks.pretty_time_string
    timecenter < 1_000.0 ? (scalefactor = 1.0; units = "ns") :
        timecenter < 1_000_000.0 ? (scalefactor = 1_000.0; units = "μs") :
            timecenter < 1_000_000_000.0 ? (scalefactor = 1_000_000.0; units = "ms") :
                (scalefactor = 1_000_000_000.0; units = " s")
    @sprintf("%6.2f%s [%6.2f,%6.2f]", timecenter/scalefactor, units, timelower/scalefactor, timeupper/scalefactor)
end

smallN, smallerN = 600, 400;
smallsqrA = sprand(smallN, smallN, 0.01);
smalltallA = sprand(smallN, smallerN, 0.01);
smallwideA = sprand(smallerN, smallN, 0.01);
largeN, largerN = 50000, 100000;
largesqrA = sprand(largerN, largerN, 0.001);
largetallA = sprand(largerN, largeN, 0.001);
largewideA = sprand(largeN, largerN, 0.001);
psmall = randperm(smallN); qsmall = randperm(smallN);
psmaller = randperm(smallerN); qsmaller = randperm(smallerN);
plarge = randperm(largeN); qlarge = randperm(largeN);
plarger = randperm(largerN); qlarger = randperm(largerN);

wm, wt = 12, 24;
println(" $(lpad("method ", wm)) $(lpad("small square A", wt)), $(lpad("small tall A", wt)), $(lpad("small wide A", wt)) | $(lpad("large square A", wt)), $(lpad("large tall A", wt)), $(lpad("large wide A", wt)) ")
@printf("%s: %s , %s , %s | %s, %s , %s\n", lpad("permute", wm),
    prettytimes(@benchmark permute(smallsqrA, psmall, qsmall)),
    prettytimes(@benchmark permute(smalltallA, psmall, qsmaller)),
    prettytimes(@benchmark permute(smallwideA, psmaller, qsmall)),
    prettytimes(@benchmark permute(largesqrA, plarger, qlarger)),
    prettytimes(@benchmark permute(largetallA, plarger, qlarge)),
    prettytimes(@benchmark permute(largewideA, plarge, qlarger)) )
@printf("%s: %s , %s , %s | %s, %s , %s\n", lpad("csc_permute", wm),
    prettytimes(@benchmark csc_permute(smallsqrA, psmall, qsmall)),
    prettytimes(@benchmark csc_permute(smalltallA, psmall, qsmaller)),
    prettytimes(@benchmark csc_permute(smallwideA, psmaller, qsmall)),
    prettytimes(@benchmark csc_permute(largesqrA, plarger, qlarger)),
    prettytimes(@benchmark csc_permute(largetallA, plarger, qlarge)),
    prettytimes(@benchmark csc_permute(largewideA, plarge, qlarger)) )
```

yield

``` julia
      method              small square A,                small tall A,              small wide A |            large square A,               large tall A,              large wide A
     permute:  59.50 μs [ 31.30, 87.71] ,  34.03 μs [ 32.35, 35.72] ,  41.43 μs [ 11.05, 71.80] |   1.27  s [  1.14,  1.39], 569.81 ms [529.28,610.35] , 567.68 ms [521.41,613.95]
 csc_permute: 107.57 μs [ 77.89,137.25] ,  69.29 μs [ 26.74,111.83] ,  74.63 μs [ 35.43,113.82] |   1.39  s [  1.27,  1.51], 639.68 ms [568.00,711.36] , 660.11 ms [585.24,734.98]
```

and `permute` requires signicantly less memory than `csc_permute`

``` julia
julia> @benchmark permute(largesqrA, plarger, qlarger)
================ Benchmark Results ========================
     Time per evaluation: 1.31 s [1.20 s, 1.42 s]
Proportion of time in GC: 9.83% [2.53%, 17.14%]
        Memory allocated: 306.57 mb
   Number of allocations: 14 allocations
       Number of samples: 6
   Number of evaluations: 6
 Time spent benchmarking: 9.70 s


julia> @benchmark csc_permute(largesqrA, plarger, qlarger)
================ Benchmark Results ========================
     Time per evaluation: 1.41 s [1.24 s, 1.58 s]
Proportion of time in GC: 12.20% [1.38%, 23.03%]
        Memory allocated: 459.87 mb
   Number of allocations: 31 allocations
       Number of samples: 5
   Number of evaluations: 5
 Time spent benchmarking: 9.03 s
```

To avoid tangling this PR and #16070, this PR does not presently remove/deprecate `csc_permute`. @KristofferC, what would you like me to do here with respect to #16070?

`symperm` remains on my radar.

KristofferC commented:
Regarding #16070, I have sort of ran out of steam on that PR for a while so you don't need to take any considerations to it.

Sacha0 commented:
> Regarding #16070, I have sort of ran out of steam on that PR for a while so you don't need to take any considerations to it.

Cheers. Shall I add a commit removing `csc_permute` then? Do I understand correctly that, given `csc_permute` is not exported, we can remove `csc_permute` without deprecation? Best!

KristofferC commented:
Best to ask @tkelman but that is my understanding yes. If it is possible and not too difficult I guess it wouldn't hurt to deprecate it and call your new permute method instead.

tkelman commented:
@nanosoldier `runbenchmarks("sparse", vs = ":master")`

tkelman commented on test/sparsedir/sparse.jl:
```diff
@@ -288,6 +288,60 @@ end
+    @test_throws ArgumentError permute!(X, A, p, q, (D = copy(C); resize!(D.nzval, nnz(A) - 1); D))
+    # Test common error checking of permute[!] methods / source-workcolptr compat
+    @test_throws DimensionMismatch permute!(A, p, q, C, Vector{eltype(A.rowval)}(length(A.colptr) - 1))
+    # Test gross functionality of [c]transpose[!]
```
overall?

Sacha0 replied:
Good call. Fixed on push. Thanks!

tkelman commented:
I just grepped all registered packages for `csc_permute` and the only hit was SuiteSparse.jl which is a copy of the code from base. @dmbates or others may have unregistered code that uses it, but I think it's fine to just delete it especially since they can easily use the package now.

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+the result `(f(A)Q)^T` (`map(f, transpose(A[:,q]))`) in `X`.
+
+`X`'s dimensions must match those of `transpose(A)` (`X.m == A.n` and `X.n == A.m`), and `X`
+must have enough storage to accomodate all allocated entries in `A` (`length(X.rowval) >= nnz(A)`
```
accommodate

Sacha0 replied:
Fixed on push. Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+"""
+function _computecolptrs_halfperm!{Tv,Ti<:Integer}(X::SparseMatrixCSC{Tv,Ti},
+        A::SparseMatrixCSC{Tv,Ti})
+    # Compute `transpose(A[:,q])`'s col counts. Store shifted forward one pos. in X.colptr.
```
spell out position, comments can violate the line length limit by a bit

Sacha0 replied:
Fixed on push. Thanks!

dmbates commented:
@tkelman IIRC I don't use `csc_permute` in packages any more and, if by chance I do, I know where to look for a replacement.

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+        A::SparseMatrixCSC{Tv,Ti}, q::AbstractVector, f)
+    @inbounds for Xi in 1:A.n
+        Aj = q[Xi]
+        for Ak in A.colptr[Aj]:(A.colptr[Aj + 1] - 1)
```
`nzrange(A, Aj)` ?

Sacha0 replied:
#16371 FTW. Fixed on push. Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+            `X.m (= $(X.m))`, must match source argument `A`'s column count, `A.n (= $(A.n))`")))
+    elseif length(X.rowval) < nnz(A)
+        throw(ArgumentError(string("the length of destination argument `X`'s `rowval` ",
+            "array, `length(X.rowval) (= $(length(X.rowval)))`, must be at least source ",
```
I think saying `>=` instead of "at least" would be a bit clearer

KristofferC replied:
`≧` ?

tkelman replied:
`\ge` will render right on lousy windows terminals, `\geqq` won't

Sacha0 replied:
I've written out "greater than or equal to" to remain prose-y (on push). Solid? Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+            X.nzval[Xk] = f(A.nzval[Ak])
+            X.colptr[Ai + 1] += 1
+        end
+    end
```
may want to make sure this returns something, otherwise I think it might be type-unstable if the loop happens to be empty

Sacha0 replied:
`return # kill potential type instability` on push. Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+end
+
+function ftranspose!{Tv,Ti}(X::SparseMatrixCSC{Tv,Ti}, A::SparseMatrixCSC{Tv,Ti}, f)
+    # Check compatability of source argument A and destination argument X
```
compatibility (5x)

Sacha0 replied:
Fixed on push. Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+column-permutation argument `q`.
+"""
+function _checkargs_sourcecompatperms_permute!{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv,Ti},
+         p::AbstractVector, q::AbstractVector)
```
should this also check that the inputs are permutations? if they have repeats or out-of-range values would bad things happen?

Sacha0 replied:
Yes, ideally we would ascertain that the permutations are in fact permutations. `isperm` is allocating and O(A.n) / O(A.m), however, so I forwent those checks.

tkelman replied:
can these segfault if they're using inbounds and get non-permutation data? maybe the work vectors would be usable for the scratch space of the `isperm` checks in some cases?

Sacha0 replied:
> can these segfault if they're using inbounds and get non-permutation data?

Good question. I imagine you could trigger a segfault by providing a 'permutation' including values outside `1:A.m` / `1:A.n`. So let's go with yes.



> maybe the work vectors would be usable for the scratch space of the isperm checks in some cases?

A good thought that. I'll have a look and respond with an update.

tkelman commented on test/sparsedir/sparse.jl:
```diff
@@ -288,6 +288,60 @@ end
+    @test_throws ArgumentError transpose!((B = similar(A); resize!(B.rowval, nnz(A) - 1); B), A)
+    @test_throws ArgumentError transpose!((B = similar(A); resize!(B.nzval, nnz(A) - 1); B), A)
+    # Test common error checking of permute[!] methods / source-perm compat
+    @test_throws DimensionMismatch Base.SparseArrays.permute(A, p[1:(end - 1)], q)
```
this shouldn't need qualifying?

Sacha0 replied:
Qualifications removed on push. Thanks!

tkelman replied:
I'm still seeing these 3

Sacha0 replied:
Ah, did you also mean the `# transpose[!]`, `# ctranspose[!]`, and `# permute[!]`?

tkelman replied:
I meant the `Base.SparseArrays.`

Sacha0 replied:
Same page now. This PR exports `permute` from `SparseArrays` but does not place a corresponding entry in `exports.jl` (see my question at the end of the fourth paragraph in this thread's first post). With this arrangement, an undefined reference error resulted when I tried calling `permute` without qualification in the tests. How should I proceed?

tkelman replied:
oh, I thought `permute` was already exported, aren't there other methods for it?

Sacha0 replied:
I thought so as well, but to my surprise `?` returns nothing.

tkelman replied:
Now it can definitely go. I think the "getindex can't know it's a permutation until we have a type for that" has informally sold over at least Keno and probably others around the Julia Computing office today.

Sacha0 replied:
Sorry, my English parser is failing: What can definitely go now? Thanks!

tkelman replied:
`Base.SparseArrays.`

Sacha0 replied:
Ah yes! Fixed. Thanks!

Sacha0 commented:
Addressed comments, rebased, and removed `csc_permute`. Thanks!

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/16931#issuecomment-226014748) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8a51dd28d4984bbfd4c042194a346ba159c4f0ac/1f8c29b_vs_2efec16/report.md). cc @jrevels

Sacha0 commented:
> possible performance regressions were detected

The one significant time regression should be unrelated or is spurious. The `[c]transpose!` methods appear to have picked up allocations. Having a look.

Sacha0 commented:
> The [c]transpose! methods appear to have picked up allocations. Having a look.

Regarding the allocation regression: I'm not certain how to proceed. Guidance would be much appreciated. Info so far: Calling `transpose!(bar, foo)` results in one allocation of 32 bytes rather than the previous version's zero allocations. Directly calling parent `ftranspose!(bar, foo, identity)` results in one allocation of 32 bytes as well, but directly calling its parent, `halfperm!(bar, foo, 1:foo.n, identity)`, results in two allocations of 32 bytes each, while directly calling `halfperm!(bar, foo, 1:foo.n)`, which should be identical to calling `halfperm!(bar, foo, 1:foo.n, identity)`, performs zero allocations. Running with `--track-allocation = user` produces no `*.mem` files.

Sacha0 commented:
Now with checks for permutation validity and associated tests. The performance impact of checking permutation validity is appreciable for small matrices, but `permute` nonetheless beats `csc_permute` handily. Along the same lines, should we strengthen the permutation vector typing, e.g. replace signatures like
```julia
permute{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv,Ti}, p::AbstractVector, q::AbstractVector)
```
with something like
```julia
permute{Tv,Ti<:Integer,Tp,Tq}(A::SparseMatrixCSC{Tv,Ti}, p::AbstractVector{Tp}, q::AbstractVector{Tq})
```
?

Beyond that, I think only the question of exporting `permute` and the allocation performance regression in `[c]transpose!` remain?

Sacha0 committed:
Refactors `[c]transpose[!]` methods operating on `SparseMatrixCSC`s in anticipation of introducing related `permute[!]` methods and taking advantage of #16371 for better structuring. Also adds several tests for these methods.

Sacha0 committed:
Introduces `permute[!]` methods operating on `SparseMatrixCSC`s, associated documentation, and tests. Exports `permute`.

Sacha0 commented:
I revised most method signatures for stronger typing, exported `permute`, and added the associated documentation to the manual. I dropped the commit removing `csc_permute` in favor of keeping those changes in #16070 / a descendant.

The `[c]transpose[!]` methods' (negligible) allocation regression naively seems a compiler issue rather than a problem with this pull request's code; I plan to open a corresponding tracking issue if / when this pull request clears but otherwise have moved on.

Barring additional commentary this pull request should be in shape. Thanks, and best!

tkelman commented:
Sorry for not making up our minds on this sooner, but isn't the copying `permute()` redundant with indexing with the permutation vector?

Sacha0 commented:
> Sorry for not making up our minds on this sooner

No need for apology! Never enough time in the day, and I imagine most relevant parties are fully occupied with JuliaCon at the moment (enjoy!).

> isn't the copying permute() redundant with indexing with the permutation vector?

Good point! `A[p,q]` accomplishes `permute(A, p, q)`, though I imagine the latter is more efficient by nature of specialization for the task. How should we proceed? Thanks!

tkelman commented:
> I imagine the latter is more efficient by nature of specialization

Is there any situation where that's the case that we couldn't fix by specializing `getindex` to call the same implementation?

Sacha0 commented:
> Is there any situation where that's the case that we couldn't fix by specializing getindex to call the same implementation?

Not sure. Foremost I am not certain how one would specialize `getindex` appropriately. Naively that specialization seemingly requires dispatch on a permutation vector type, perhaps an annotation type akin to `Symmetric` or `LowerTriangular`. Alternatively one could perform runtime checks in the methods for `A[x, y]`, `A[x,:]`, and `A[:,y]` that determine whether `x`/`y`are permutations and branch to the specialized implementation. Thoughts?

tkelman commented:
Ah, good point. Given that we don't have a specific type for permutation vectors in Base right now, maybe it is worth keeping the copying version as separate. I'll let others opine.

Sacha0 commented:
> we don't have a specific type for permutation vectors in Base

Does discussion of this pos

tkelman commented:
There used to be a Permutations type, but that was moved to Combinatorics and I don't think it had the same meaning of a permutation vector that we're referring to here.

Sacha0 commented:
@tkelman, when / at what interval are bumps generally (i.e. outside of a nearing-release period) appreciated? In a nearing-release period, should bumps be held till after the release? Thanks and best!

tkelman commented:
Somewhere between a few days to a week I guess? We're sort of in an unofficial feature freeze already while we finish up the last few items on the milestone, but this was pretty much ready as far as I can tell so I think I'll merge it tomorrow if no one says otherwise. Thanks!

Sacha0 commented:
Cheers, I'll hold off with other bumps till the 0.6 dev cycle. Let me know if I can lend a hand with anything prior. Thanks again and best!

ViralBShah commented:
Would be nice to merge this. Should not interact with much of anything.

Sacha0 commented:
Thanks for the review and merge!
